File: apps/web/src/lib/trpc/statuses.ts
// EXP-314 — per-team custom issue statuses (Linear-style). Member-managed
// like labels (`mutate_resources`, no owner gate — the user decision for this
// feature); the 7 builtin rows (builtin_key != NULL) are locked: never
// renamed, recolored, or deleted, only moved within their category. Category
// writes serialize on a FOR UPDATE lock over the category's rows so the
// started cap (ISSUE_STATUS_STARTED_MAX) can't be raced past and concurrent
// moves keep a consistent order.
import { z } from "zod"
import { and, asc, eq, ne, sql } from "drizzle-orm"
import { TRPCError } from "@trpc/server"
import { router, authedProcedure, generateTxId } from "@/lib/trpc"
import { issues, issueStatuses, teams } from "@/db/schema"
import { resolveTeamAccess } from "@/lib/team-membership"
import {
  CATEGORY_ANCHOR,
  hexColorSchema,
  ISSUE_STATUS_STARTED_MAX,
  type IssueStatusCategory,
  issueStatusCategorySchema,
} from "@/lib/domain"
import { applyStatusDerivations } from "@/lib/status-derivations"
import { recordIssueEvent } from "@/lib/integrations/activity"
import { isUniqueViolation } from "@/lib/trpc/db-errors"

const statusNameSchema = z.string().min(1).max(255)
const statusColorSchema = hexColorSchema

function duplicateNameError(name: string): TRPCError {
  return new TRPCError({
    code: `CONFLICT`,
    message: `A status named "${name}" already exists in this team`,
  })
}

type Tx = Parameters<
  // eslint-disable-next-line quotes
  Parameters<typeof import("@/db/connection").db.transaction>[0]
>[0]

// Serialize ALL structural status writes for a team on one advisory lock,
// taken as the FIRST statement of every create/move/delete transaction.
// Row locks alone are not enough: FOR UPDATE cannot block phantom INSERTs
// (two concurrent creates each counting 3 started rows would both pass the
// cap), and move/delete's row-then-category lock order could deadlock
// against create's category-first order. xact-scoped — released at commit.
async function lockTeamStatuses(tx: Tx, teamId: string): Promise<void> {
  await tx.execute(
    sql`SELECT pg_advisory_xact_lock(hashtext(${`issue_statuses:${teamId}`}))`
  )
}

// Lock the category's rows in display order — the per-row serialization for
// the started cap, appends, and moves (lockTeamStatuses guards the phantoms).
async function lockCategory(
  tx: Tx,
  teamId: string,
  category: IssueStatusCategory
) {
  return await tx
    .select({
      id: issueStatuses.id,
      sortOrder: issueStatuses.sortOrder,
    })
    .from(issueStatuses)
    .where(
      and(
        eq(issueStatuses.teamId, teamId),
        eq(issueStatuses.category, category)
      )
    )
    .orderBy(
      asc(issueStatuses.sortOrder),
      asc(issueStatuses.createdAt),
      asc(issueStatuses.id)
    )
    .for(`update`)
}

async function loadStatusForWrite(tx: Tx, teamId: string, statusId: string) {
  const [row] = await tx
    .select()
    .from(issueStatuses)
    .where(
      and(eq(issueStatuses.id, statusId), eq(issueStatuses.teamId, teamId))
    )
    .limit(1)
    .for(`update`)
  if (!row) {
    throw new TRPCError({ code: `NOT_FOUND`, message: `Status not found` })
  }
  return row
}

function assertNotBuiltin(row: { builtinKey: string | null }): void {
  if (row.builtinKey !== null) {
    throw new TRPCError({
      code: `BAD_REQUEST`,
      message: `Built-in statuses cannot be edited or deleted`,
    })
  }
}

export const statusesRouter = router({
  // Server-authoritative count of issues still referencing a status. The
  // FK doesn't care about board trash, so this includes issues on trashed
  // boards — which never sync, so the client's synced count can undershoot.
  // Powers the delete dialog's "N issues will move to …" copy (EXP-320).
  referencingCount: authedProcedure
    .input(
      z.object({
        teamId: z.string().uuid(),
        statusId: z.string().uuid(),
      })
    )
    .query(async ({ ctx, input }) => {
      await resolveTeamAccess(ctx.session.user.id, input.teamId, `read`)
      const [status] = await ctx.db
        .select({ id: issueStatuses.id })
        .from(issueStatuses)
        .where(
          and(
            eq(issueStatuses.id, input.statusId),
            eq(issueStatuses.teamId, input.teamId)
          )
        )
        .limit(1)
      if (!status) {
        throw new TRPCError({ code: `NOT_FOUND`, message: `Status not found` })
      }
      const [row] = await ctx.db
        .select({ count: sql<number>`count(*)::int` })
        .from(issues)
        .where(eq(issues.statusId, input.statusId))
      return { count: row?.count ?? 0 }
    }),

  create: authedProcedure
    .input(
      z.object({
        teamId: z.string().uuid(),
        category: issueStatusCategorySchema,
        name: statusNameSchema,
        color: statusColorSchema,
      })
    )
    .mutation(async ({ ctx, input }) => {
      await resolveTeamAccess(
        ctx.session.user.id,
        input.teamId,
        `mutate_resources`
      )
      if (input.category === `duplicate`) {
        // Duplicate is a fixed single-status category (no + button anywhere).
        throw new TRPCError({
          code: `BAD_REQUEST`,
          message: `The duplicate category cannot have custom statuses`,
        })
      }
      try {
        return await ctx.db.transaction(async (tx) => {
          const txId = await generateTxId(tx)
          await lockTeamStatuses(tx, input.teamId)
          const siblings = await lockCategory(tx, input.teamId, input.category)
          if (
            input.category === `started` &&
            siblings.length >= ISSUE_STATUS_STARTED_MAX
          ) {
            // The pie-clock fill tables are defined only up to 4 started
            // statuses.
            throw new TRPCError({
              code: `PRECONDITION_FAILED`,
              message: `A team can have at most ${ISSUE_STATUS_STARTED_MAX} started statuses`,
            })
          }

          // Pre-check against (team_id, lower(name)) so the caller gets a
          // readable CONFLICT instead of a raw 23505 (labels EXP-254
          // convention). Also blocks shadowing a builtin's name — intended.
          const [clash] = await tx
            .select({ id: issueStatuses.id })
            .from(issueStatuses)
            .where(
              and(
                eq(issueStatuses.teamId, input.teamId),
                sql`lower(${issueStatuses.name}) = lower(${input.name})`
              )
            )
            .limit(1)
          if (clash) throw duplicateNameError(input.name)

          const maxSort = siblings.reduce(
            (max, row) => Math.max(max, row.sortOrder),
            0
          )
          const [status] = await tx
            .insert(issueStatuses)
            .values({
              teamId: input.teamId,
              category: input.category,
              name: input.name,
              color: input.color,
              sortOrder: maxSort + 1,
              builtinKey: null,
            })
            .returning()

          return { txId, status }
        })
      } catch (err) {
        // The pre-check races concurrent creators — translate a late unique
        // violation into the same CONFLICT.
        if (isUniqueViolation(err)) throw duplicateNameError(input.name)
        throw err
      }
    }),

  update: authedProcedure
    .input(
      z.object({
        teamId: z.string().uuid(),
        statusId: z.string().uuid(),
        name: statusNameSchema.optional(),
        color: statusColorSchema.optional(),
        // Deliberately NO category input — a status's category is immutable
        // (moving an issue population between lifecycle semantics silently
        // would corrupt completedAt/anchor invariants).
      })
    )
    .mutation(async ({ ctx, input }) => {
      await resolveTeamAccess(
        ctx.session.user.id,
        input.teamId,
        `mutate_resources`
      )
      try {
        return await ctx.db.transaction(async (tx) => {
          const txId = await generateTxId(tx)
          const row = await loadStatusForWrite(tx, input.teamId, input.statusId)
          assertNotBuiltin(row)

          const updates: { name?: string; color?: string } = {}
          if (input.name !== undefined) updates.name = input.name
          if (input.color !== undefined) updates.color = input.color

          if (updates.name !== undefined) {
            const [clash] = await tx
              .select({ id: issueStatuses.id })
              .from(issueStatuses)
              .where(
                and(
                  eq(issueStatuses.teamId, input.teamId),
                  ne(issueStatuses.id, input.statusId),
                  sql`lower(${issueStatuses.name}) = lower(${updates.name})`
                )
              )
              .limit(1)
            if (clash) throw duplicateNameError(updates.name)
          }

          // EXP-707: echo the row (envelope rule: mutations return
          // { row, txId }) so callers can read back what they wrote.
          let status = row
          if (Object.keys(updates).length > 0) {
            ;[status] = await tx
              .update(issueStatuses)
              .set(updates)
              .where(eq(issueStatuses.id, input.statusId))
              .returning()
          }

          return { txId, status }
        })
      } catch (err) {
        if (isUniqueViolation(err) && input.name !== undefined) {
          throw duplicateNameError(input.name)
        }
        throw err
      }
    }),

  // Swap a status with its neighbor within its category. Builtins ARE movable
  // (only name/color/delete are locked): reordering started statuses is how a
  // team shapes the clock fills. The whole category is re-indexed 1..n first,
  // so historic sort_order ties can never make a move a silent no-op.
  move: authedProcedure
    .input(
      z.object({
        teamId: z.string().uuid(),
        statusId: z.string().uuid(),
        direction: z.enum([`up`, `down`]),
      })
    )
    .mutation(async ({ ctx, input }) => {
      await resolveTeamAccess(
        ctx.session.user.id,
        input.teamId,
        `mutate_resources`
      )
      return await ctx.db.transaction(async (tx) => {
        const txId = await generateTxId(tx)
        await lockTeamStatuses(tx, input.teamId)
        const row = await loadStatusForWrite(tx, input.teamId, input.statusId)
        const siblings = await lockCategory(tx, input.teamId, row.category)

        const order = siblings.map((s) => s.id)
        const index = order.indexOf(input.statusId)
        const targetIndex = input.direction === `up` ? index - 1 : index + 1
        if (targetIndex >= 0 && targetIndex < order.length) {
          ;[order[index], order[targetIndex]] = [
            order[targetIndex],
            order[index],
          ]
        }
        // Persist positions 1..n, touching only rows whose position changed
        // (an edge move is an idempotent no-op — safe for double-clicks).
        for (let position = 0; position < order.length; position++) {
          const sibling = siblings.find((s) => s.id === order[position])!
          if (sibling.sortOrder !== position + 1) {
            await tx
              .update(issueStatuses)
              .set({ sortOrder: position + 1 })
              .where(eq(issueStatuses.id, sibling.id))
          }
        }

        return { txId }
      })
    }),

  delete: authedProcedure
    .input(
      z.object({
        teamId: z.string().uuid(),
        statusId: z.string().uuid(),
        // Where this status's issues go. Required whenever any issue still
        // references the status (the count includes trashed-board issues —
        // the FK doesn't care about trash, so the client's synced count can
        // undershoot; always honor the PRECONDITION_FAILED).
        reassignToId: z.string().uuid().optional(),
      })
    )
    .mutation(async ({ ctx, input }) => {
      await resolveTeamAccess(
        ctx.session.user.id,
        input.teamId,
        `mutate_resources`
      )
      return await ctx.db.transaction(async (tx) => {
        const txId = await generateTxId(tx)
        await lockTeamStatuses(tx, input.teamId)
        const row = await loadStatusForWrite(tx, input.teamId, input.statusId)
        assertNotBuiltin(row)

        // Lock every referencing issue (ordered like bulkUpdate for deadlock
        // hygiene) — the reassignment below derives per-issue values.
        const referencing = await tx
          .select({
            id: issues.id,
            status: issues.status,
            duplicateOfId: issues.duplicateOfId,
          })
          .from(issues)
          .where(eq(issues.statusId, input.statusId))
          .orderBy(issues.id)
          .for(`update`)

        let target: typeof issueStatuses.$inferSelect | null = null
        if (referencing.length > 0) {
          if (!input.reassignToId) {
            throw new TRPCError({
              code: `PRECONDITION_FAILED`,
              message: `${referencing.length} issues use this status. Pick a replacement first.`,
            })
          }
          if (input.reassignToId === input.statusId) {
            throw new TRPCError({
              code: `BAD_REQUEST`,
              message: `Cannot reassign issues to the status being deleted`,
            })
          }
          const [targetRow] = await tx
            .select()
            .from(issueStatuses)
            .where(
              and(
                eq(issueStatuses.id, input.reassignToId),
                eq(issueStatuses.teamId, input.teamId)
              )
            )
            .limit(1)
          if (!targetRow) {
            throw new TRPCError({
              code: `BAD_REQUEST`,
              message: `Replacement status not found`,
            })
          }
          if (targetRow.category === `duplicate`) {
            throw new TRPCError({
              code: `BAD_REQUEST`,
              message: `Issues cannot be reassigned to the duplicate status`,
            })
          }
          target = targetRow

          const targetAnchor =
            targetRow.builtinKey ?? CATEGORY_ANCHOR[targetRow.category]
          for (const issue of referencing) {
            const setValues: Record<string, unknown> = {
              status: targetAnchor,
              statusId: targetRow.id,
            }
            // Same completedAt/duplicate rules as any status write.
            applyStatusDerivations(setValues, issue)
            await tx
              .update(issues)
              .set(setValues)
              .where(eq(issues.id, issue.id))
            await recordIssueEvent(tx, {
              issueId: issue.id,
              teamId: input.teamId,
              actorUserId: ctx.session.user.id,
              type: `status_changed`,
              payload: {
                fromStatusId: row.id,
                toStatusId: targetRow.id,
                fromName: row.name,
                toName: targetRow.name,
              },
            })
          }
          // Deliberately NO notification fan-out: deleting a status is an
          // admin bulk action — pinging every subscriber of every reassigned
          // issue would be pure noise (mirrors bulkUpdate's fan-out cap
          // rationale, taken to zero).
        }

        await tx
          .delete(issueStatuses)
          .where(eq(issueStatuses.id, input.statusId))

        return {
          txId,
          reassigned: referencing.length,
          reassignedToId: target?.id ?? null,
        }
      })
    }),

  // EXP-319 — per-team PR automation targets ("when a PR opens/merges, move
  // issues to …"). Member-gated like every other write in this router: the
  // setting is part of status management, not team administration. Writes
  // the teams columns applyPrLifecycleStatusInTx (pr-sync) reads: a status
  // row id pins the target, 'default' resets to the builtin (NULL — also
  // what the FK's SET NULL leaves behind when a target status is deleted),
  // 'none' disables the automation entirely ("do nothing" — merged PRs then
  // never auto-complete their issues).
  setPrAutomation: authedProcedure
    .input(
      z.object({
        teamId: z.string().uuid(),
        event: z.enum([`pr_opened`, `pr_merged`]),
        target: z.union([
          z.string().uuid(),
          z.literal(`default`),
          z.literal(`none`),
        ]),
      })
    )
    .mutation(async ({ ctx, input }) => {
      await resolveTeamAccess(
        ctx.session.user.id,
        input.teamId,
        `mutate_resources`
      )
      return await ctx.db.transaction(async (tx) => {
        const txId = await generateTxId(tx)

        let statusId: string | null = null
        if (input.target !== `default` && input.target !== `none`) {
          const [row] = await tx
            .select({
              id: issueStatuses.id,
              category: issueStatuses.category,
            })
            .from(issueStatuses)
            .where(
              and(
                eq(issueStatuses.id, input.target),
                eq(issueStatuses.teamId, input.teamId)
              )
            )
            .limit(1)
          if (!row) {
            throw new TRPCError({
              code: `NOT_FOUND`,
              message: `Status not found`,
            })
          }
          if (row.category === `duplicate`) {
            // Duplicate requires a canonical issue (enum + duplicateOfId
            // lockstep) — an automation can never supply one.
            throw new TRPCError({
              code: `BAD_REQUEST`,
              message: `Duplicate cannot be an automation target`,
            })
          }
          statusId = row.id
        }

        await tx
          .update(teams)
          .set(
            input.event === `pr_opened`
              ? {
                  prOpenedStatusId: statusId,
                  prOpenedAutomation: input.target !== `none`,
                }
              : {
                  prMergedStatusId: statusId,
                  prMergedAutomation: input.target !== `none`,
                }
          )
          .where(eq(teams.id, input.teamId))

        return { txId }
      })
    }),

  // EXP-711 — does a merged PR end the live coding sessions on its issues?
  // Member-gated like setPrAutomation (same card, same trust). The merging
  // run's own spare (`merged_own_pr`) and MCP `pr_merge`'s per-call
  // `endSessions` override are unaffected by this default.
  setEndSessionsOnMerge: authedProcedure
    .input(
      z.object({
        teamId: z.string().uuid(),
        enabled: z.boolean(),
      })
    )
    .mutation(async ({ ctx, input }) => {
      await resolveTeamAccess(
        ctx.session.user.id,
        input.teamId,
        `mutate_resources`
      )
      return await ctx.db.transaction(async (tx) => {
        const txId = await generateTxId(tx)
        await tx
          .update(teams)
          .set({ endSessionsOnMerge: input.enabled })
          .where(eq(teams.id, input.teamId))
        return { txId }
      })
    }),
})

File: apps/desktop/crates/api/src/statuses.rs
//! Typed `statuses.*` tRPC helpers (EXP-314 — Settings → Issue statuses).
//! Verified against `apps/web/src/lib/trpc/statuses.ts`:
//!
//! - `statuses.create({teamId, category, name, color})` → `{txId, status}`
//! - `statuses.update({teamId, statusId, name?, color?})` → `{txId}` —
//!   builtins are locked server-side (`BAD_REQUEST`); there is deliberately NO
//!   `category` input (a status's category is immutable)
//! - `statuses.move({teamId, statusId, direction})` → `{txId}` — swaps with
//!   the neighbor WITHIN the category; builtins ARE movable
//! - `statuses.delete({teamId, statusId, reassignToId?})` →
//!   `{txId, reassigned, reassignedToId}` — `reassignToId` is REQUIRED
//!   whenever issues still reference the status (`PRECONDITION_FAILED`
//!   otherwise; the server's count includes trashed-board issues, so the
//!   client's synced count can undershoot — always honor the error)
//! - `statuses.referencingCount({teamId, statusId})` → `{count}` — query
//!   (EXP-320): the server-authoritative referencing-issue count, trashed
//!   boards included — the delete dialog's copy source
//!
//! Writes are member-level (`mutate_resources`, like labels — no owner gate).
//! Reads come from the synced `issue_statuses` collection, never a tRPC list
//! call (§4.1). All calls are blocking; background executor only (§3.5).

use serde::{Deserialize, Serialize};

use crate::error::ApiError;
use crate::labels::TxOutput;
use crate::trpc::TrpcClient;

/// Slim camelCase mirror of the row `statuses.create` returns.
#[derive(Clone, Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StatusOut {
    pub id: String,
    #[serde(default)]
    pub team_id: Option<String>,
    #[serde(default)]
    pub category: Option<String>,
    #[serde(default)]
    pub name: Option<String>,
    #[serde(default)]
    pub color: Option<String>,
    #[serde(default)]
    pub builtin_key: Option<String>,
}

#[derive(Clone, Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StatusesCreateOutput {
    pub status: StatusOut,
    #[serde(default)]
    pub tx_id: Option<i64>,
}

/// Output of `statuses.delete`: how many issues were moved, and where.
#[derive(Clone, Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StatusesDeleteOutput {
    #[serde(default)]
    pub reassigned: Option<i64>,
    #[serde(default)]
    pub reassigned_to_id: Option<String>,
    #[serde(default)]
    pub tx_id: Option<i64>,
}

/// Which way [`statuses_move`] shifts a status within its category.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum MoveDirection {
    Up,
    Down,
}

impl MoveDirection {
    fn as_wire(self) -> &'static str {
        match self {
            MoveDirection::Up => "up",
            MoveDirection::Down => "down",
        }
    }
}

/// `statuses.create` — mutation. `category` is a
/// `domain::contract::ISSUE_STATUS_CATEGORY_VALUES` wire value (never
/// `duplicate` — the server refuses); `color` is a required `#rrggbb`.
pub fn statuses_create(
    trpc: &TrpcClient,
    team_id: &str,
    category: &str,
    name: &str,
    color: &str,
) -> Result<StatusesCreateOutput, ApiError> {
    #[derive(Serialize)]
    #[serde(rename_all = "camelCase")]
    struct Input<'a> {
        team_id: &'a str,
        category: &'a str,
        name: &'a str,
        color: &'a str,
    }
    trpc.mutation(
        "statuses.create",
        &Input {
            team_id,
            category,
            name,
            color,
        },
    )
}

/// `statuses.update` — mutation (inline name/color edits on CUSTOM rows).
pub fn statuses_update(
    trpc: &TrpcClient,
    team_id: &str,
    status_id: &str,
    name: Option<&str>,
    color: Option<&str>,
) -> Result<TxOutput, ApiError> {
    #[derive(Serialize)]
    #[serde(rename_all = "camelCase")]
    struct Input<'a> {
        team_id: &'a str,
        status_id: &'a str,
        #[serde(skip_serializing_if = "Option::is_none")]
        name: Option<&'a str>,
        #[serde(skip_serializing_if = "Option::is_none")]
        color: Option<&'a str>,
    }
    trpc.mutation(
        "statuses.update",
        &Input {
            team_id,
            status_id,
            name,
            color,
        },
    )
}

/// `statuses.move` — mutation. A move past a category edge is an idempotent
/// server-side no-op (safe for double-clicks).
pub fn statuses_move(
    trpc: &TrpcClient,
    team_id: &str,
    status_id: &str,
    direction: MoveDirection,
) -> Result<TxOutput, ApiError> {
    #[derive(Serialize)]
    #[serde(rename_all = "camelCase")]
    struct Input<'a> {
        team_id: &'a str,
        status_id: &'a str,
        direction: &'a str,
    }
    trpc.mutation(
        "statuses.move",
        &Input {
            team_id,
            status_id,
            direction: direction.as_wire(),
        },
    )
}

/// Output of `statuses.referencingCount`: the server-authoritative number of
/// issues still referencing a status (trashed-board issues included).
#[derive(Clone, Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StatusesReferencingCountOutput {
    pub count: i64,
}

/// `statuses.referencingCount` — query (EXP-320). The synced issue count can
/// undershoot the server's (trashed-board issues never sync); the delete
/// dialog shows this count instead.
pub fn statuses_referencing_count(
    trpc: &TrpcClient,
    team_id: &str,
    status_id: &str,
) -> Result<StatusesReferencingCountOutput, ApiError> {
    #[derive(Serialize)]
    #[serde(rename_all = "camelCase")]
    struct Input<'a> {
        team_id: &'a str,
        status_id: &'a str,
    }
    trpc.query_with_input(
        "statuses.referencingCount",
        &Input { team_id, status_id },
    )
}

/// `statuses.delete` — mutation. Pass `reassign_to_id` whenever any issue
/// still sits in the status; omitting it there fails with a
/// `PRECONDITION_FAILED` naming the count.
pub fn statuses_delete(
    trpc: &TrpcClient,
    team_id: &str,
    status_id: &str,
    reassign_to_id: Option<&str>,
) -> Result<StatusesDeleteOutput, ApiError> {
    #[derive(Serialize)]
    #[serde(rename_all = "camelCase")]
    struct Input<'a> {
        team_id: &'a str,
        status_id: &'a str,
        #[serde(skip_serializing_if = "Option::is_none")]
        reassign_to_id: Option<&'a str>,
    }
    trpc.mutation(
        "statuses.delete",
        &Input {
            team_id,
            status_id,
            reassign_to_id,
        },
    )
}

/// Which PR event [`statuses_set_pr_automation`] configures (EXP-319).
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum PrAutomationEvent {
    Opened,
    Merged,
}

impl PrAutomationEvent {
    fn as_wire(self) -> &'static str {
        match self {
            PrAutomationEvent::Opened => "pr_opened",
            PrAutomationEvent::Merged => "pr_merged",
        }
    }
}

/// The target [`statuses_set_pr_automation`] writes: a status row uuid,
/// `Default` (reset to NULL — the builtin In Review/Done fallback), or
/// `DoNothing` (disable the automation).
#[derive(Clone, Debug, PartialEq, Eq)]
pub enum PrAutomationTarget {
    Status(String),
    Default,
    DoNothing,
}

impl PrAutomationTarget {
    fn as_wire(&self) -> &str {
        match self {
            PrAutomationTarget::Status(id) => id,
            PrAutomationTarget::Default => "default",
            PrAutomationTarget::DoNothing => "none",
        }
    }
}

/// `statuses.setPrAutomation` — mutation (EXP-319, member-level like every
/// other write in this file). Sets where issues move when a PR opens/merges;
/// convergence is the teams-shape Electric echo.
pub fn statuses_set_pr_automation(
    trpc: &TrpcClient,
    team_id: &str,
    event: PrAutomationEvent,
    target: &PrAutomationTarget,
) -> Result<TxOutput, ApiError> {
    #[derive(Serialize)]
    #[serde(rename_all = "camelCase")]
    struct Input<'a> {
        team_id: &'a str,
        event: &'a str,
        target: &'a str,
    }
    trpc.mutation(
        "statuses.setPrAutomation",
        &Input {
            team_id,
            event: event.as_wire(),
            target: target.as_wire(),
        },
    )
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::trpc::tests::one_shot_server;
    use crate::StaticToken;
    use std::sync::Arc;
    use std::time::Duration;

    fn client(base: &str) -> TrpcClient {
        TrpcClient::new(base, Arc::new(StaticToken("tok".to_string())))
    }

    #[test]
    fn create_posts_the_full_camel_case_body() {
        let (base, captured) = one_shot_server(
            200,
            r##"{"result":{"data":{"txId":5,"status":{"id":"s-1","teamId":"t-1","category":"started","name":"QA","color":"#22c55e","builtinKey":null}}}}"##,
        );
        let out = statuses_create(&client(&base), "t-1", "started", "QA", "#22c55e").unwrap();
        assert_eq!(out.status.name.as_deref(), Some("QA"));
        assert_eq!(out.status.builtin_key, None);
        assert_eq!(out.tx_id, Some(5));
        let request = captured.recv_timeout(Duration::from_secs(5)).unwrap();
        assert!(request.starts_with("POST /api/trpc/statuses.create HTTP/1.1"));
        assert!(request
            .ends_with(r##"{"teamId":"t-1","category":"started","name":"QA","color":"#22c55e"}"##));
    }

    #[test]
    fn update_omits_untouched_fields() {
        let (base, captured) = one_shot_server(200, r#"{"result":{"data":{"txId":6}}}"#);
        let out = statuses_update(&client(&base), "t-1", "s-1", Some("Testing"), None).unwrap();
        assert_eq!(out.tx_id, Some(6));
        let request = captured.recv_timeout(Duration::from_secs(5)).unwrap();
        assert!(request.ends_with(r#"{"teamId":"t-1","statusId":"s-1","name":"Testing"}"#));

        let (base, captured) = one_shot_server(200, r#"{"result":{"data":{"txId":7}}}"#);
        let _ = statuses_update(&client(&base), "t-1", "s-1", None, Some("#ef4444")).unwrap();
        let request = captured.recv_timeout(Duration::from_secs(5)).unwrap();
        assert!(request.ends_with(r##"{"teamId":"t-1","statusId":"s-1","color":"#ef4444"}"##));
    }

    #[test]
    fn move_sends_the_direction_wire_value() {
        let (base, captured) = one_shot_server(200, r#"{"result":{"data":{"txId":8}}}"#);
        let _ = statuses_move(&client(&base), "t-1", "s-1", MoveDirection::Up).unwrap();
        let request = captured.recv_timeout(Duration::from_secs(5)).unwrap();
        assert!(request.starts_with("POST /api/trpc/statuses.move HTTP/1.1"));
        assert!(request.ends_with(r#"{"teamId":"t-1","statusId":"s-1","direction":"up"}"#));

        let (base, captured) = one_shot_server(200, r#"{"result":{"data":{"txId":9}}}"#);
        let _ = statuses_move(&client(&base), "t-1", "s-1", MoveDirection::Down).unwrap();
        let request = captured.recv_timeout(Duration::from_secs(5)).unwrap();
        assert!(request.ends_with(r#"{"teamId":"t-1","statusId":"s-1","direction":"down"}"#));
    }

    #[test]
    fn set_pr_automation_sends_all_three_target_forms() {
        let (base, captured) = one_shot_server(200, r#"{"result":{"data":{"txId":12}}}"#);
        let out = statuses_set_pr_automation(
            &client(&base),
            "t-1",
            PrAutomationEvent::Opened,
            &PrAutomationTarget::Status("s-1".to_string()),
        )
        .unwrap();
        assert_eq!(out.tx_id, Some(12));
        let request = captured.recv_timeout(Duration::from_secs(5)).unwrap();
        assert!(request.starts_with("POST /api/trpc/statuses.setPrAutomation HTTP/1.1"));
        assert!(request.ends_with(r#"{"teamId":"t-1","event":"pr_opened","target":"s-1"}"#));

        let (base, captured) = one_shot_server(200, r#"{"result":{"data":{"txId":13}}}"#);
        let _ = statuses_set_pr_automation(
            &client(&base),
            "t-1",
            PrAutomationEvent::Merged,
            &PrAutomationTarget::DoNothing,
        )
        .unwrap();
        let request = captured.recv_timeout(Duration::from_secs(5)).unwrap();
        assert!(request.ends_with(r#"{"teamId":"t-1","event":"pr_merged","target":"none"}"#));

        let (base, captured) = one_shot_server(200, r#"{"result":{"data":{"txId":14}}}"#);
        let _ = statuses_set_pr_automation(
            &client(&base),
            "t-1",
            PrAutomationEvent::Merged,
            &PrAutomationTarget::Default,
        )
        .unwrap();
        let request = captured.recv_timeout(Duration::from_secs(5)).unwrap();
        assert!(request.ends_with(r#"{"teamId":"t-1","event":"pr_merged","target":"default"}"#));
    }

    #[test]
    fn delete_omits_the_reassign_target_when_none() {
        let (base, captured) = one_shot_server(
            200,
            r#"{"result":{"data":{"txId":10,"reassigned":0,"reassignedToId":null}}}"#,
        );
        let out = statuses_delete(&client(&base), "t-1", "s-1", None).unwrap();
        assert_eq!(out.reassigned, Some(0));
        assert_eq!(out.reassigned_to_id, None);
        let request = captured.recv_timeout(Duration::from_secs(5)).unwrap();
        assert!(request.ends_with(r#"{"teamId":"t-1","statusId":"s-1"}"#));

        let (base, captured) = one_shot_server(
            200,
            r#"{"result":{"data":{"txId":11,"reassigned":3,"reassignedToId":"s-2"}}}"#,
        );
        let out = statuses_delete(&client(&base), "t-1", "s-1", Some("s-2")).unwrap();
        assert_eq!(out.reassigned, Some(3));
        assert_eq!(out.reassigned_to_id.as_deref(), Some("s-2"));
        let request = captured.recv_timeout(Duration::from_secs(5)).unwrap();
        assert!(
            request.ends_with(r#"{"teamId":"t-1","statusId":"s-1","reassignToId":"s-2"}"#)
        );
    }

    #[test]
    fn referencing_count_is_a_get_query() {
        let (base, captured) = one_shot_server(200, r#"{"result":{"data":{"count":3}}}"#);
        let out = statuses_referencing_count(&client(&base), "t-1", "s-1").unwrap();
        assert_eq!(out.count, 3);
        let request = captured.recv_timeout(Duration::from_secs(5)).unwrap();
        assert!(request.starts_with("GET /api/trpc/statuses.referencingCount?input="));
    }

    #[test]
    fn precondition_failures_surface_their_message() {
        // The reassign-required refusal the pane renders inline.
        let (base, _captured) = one_shot_server(
            412,
            r#"{"error":{"json":{"message":"3 issues use this status. Pick a replacement first.","code":-32003}}}"#,
        );
        let err = statuses_delete(&client(&base), "t-1", "s-1", None).unwrap_err();
        match err {
            ApiError::Http { message, .. } => {
                assert!(message.contains("Pick a replacement first"), "{message}");
            }
            other => panic!("expected an HTTP error, got {other:?}"),
        }
    }
}
